File: src/lib/parser.ts
import Parser from "rss-parser";

const parser = new Parser({
  timeout: 15000,
  headers: {
    "User-Agent": "Everything-RSS/1.0",
  },
  customFields: {
    item: [
      ["media:content", "mediaContent", { keepArray: false }],
      ["media:thumbnail", "mediaThumbnail", { keepArray: false }],
      ["enclosure", "enclosure", { keepArray: false }],
    ],
  },
});

export interface ParsedEntry {
  guid: string;
  title: string;
  url: string;
  content: string;
  summary: string;
  author: string;
  thumbnail: string | null;
  mediaUrl: string | null;
  publishedAt: number | null;
}

export interface ParsedFeed {
  title: string;
  siteUrl: string;
  description: string;
  entries: ParsedEntry[];
}

function extractThumbnail(item: Record<string, unknown>): string | null {
  const mediaThumbnail = item.mediaThumbnail as Record<string, unknown> | undefined;
  if (mediaThumbnail?.$ && (mediaThumbnail.$ as Record<string, string>).url) {
    return (mediaThumbnail.$ as Record<string, string>).url;
  }
  const mediaContent = item.mediaContent as Record<string, unknown> | undefined;
  if (mediaContent?.$ && (mediaContent.$ as Record<string, string>).url) {
    const type = (mediaContent.$ as Record<string, string>).type || "";
    if (type.startsWith("image/")) {
      return (mediaContent.$ as Record<string, string>).url;
    }
  }
  return null;
}

function extractMediaUrl(item: Record<string, unknown>): string | null {
  const mediaContent = item.mediaContent as Record<string, unknown> | undefined;
  if (mediaContent?.$ && (mediaContent.$ as Record<string, string>).url) {
    const type = (mediaContent.$ as Record<string, string>).type || "";
    if (type.startsWith("video/") || type.startsWith("audio/")) {
      return (mediaContent.$ as Record<string, string>).url;
    }
  }
  const enclosure = item.enclosure as Record<string, unknown> | undefined;
  if (enclosure?.url) {
    return enclosure.url as string;
  }
  return null;
}

export async function parseFeed(url: string): Promise<ParsedFeed> {
  const feed = await parser.parseURL(url);

  const entries: ParsedEntry[] = (feed.items || []).map((item) => {
    const raw = item as unknown as Record<string, unknown>;
    return {
      guid: (item.guid || raw.id || item.link || item.title || "") as string,
      title: (item.title || "") as string,
      url: (item.link || "") as string,
      content: ((item.content || raw["content:encoded"] || "") as string),
      summary: (item.contentSnippet || raw.summary || "") as string,
      author: (item.creator || raw.author || "") as string,
      thumbnail: extractThumbnail(raw),
      mediaUrl: extractMediaUrl(raw),
      publishedAt: item.isoDate ? Math.floor(new Date(item.isoDate).getTime() / 1000) : null,
    };
  });

  return {
    title: feed.title || "",
    siteUrl: feed.link || "",
    description: feed.description || "",
    entries,
  };
}

File: backend/internal/fetcher/parser.go
package fetcher

import (
	"net/http"
	"time"

	"github.com/google/uuid"
	"github.com/mmcdole/gofeed"
	"github.com/qing-turnaround/everything-rss/backend/internal/model"
)

func ParseFeed(url string) ([]model.Entry, error) {
	fp := gofeed.NewParser()
	fp.UserAgent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"
	fp.Client = &http.Client{Timeout: 30 * time.Second}
	feed, err := fp.ParseURL(url)
	if err != nil {
		return nil, err
	}

	var entries []model.Entry
	for _, item := range feed.Items {
		e := model.Entry{
			ID:   uuid.New().String(),
			GUID: itemGUID(item),
		}

		if item.Title != "" {
			e.Title = &item.Title
		}
		if item.Link != "" {
			e.URL = &item.Link
		}
		if item.Content != "" {
			e.Content = &item.Content
		}
		if item.Description != "" {
			e.Summary = &item.Description
		}
		if len(item.Authors) > 0 && item.Authors[0].Name != "" {
			e.Author = &item.Authors[0].Name
		}

		if item.PublishedParsed != nil {
			ts := item.PublishedParsed.Unix()
			e.PublishedAt = &ts
		} else if item.UpdatedParsed != nil {
			ts := item.UpdatedParsed.Unix()
			e.PublishedAt = &ts
		} else {
			ts := time.Now().Unix()
			e.PublishedAt = &ts
		}

		thumb, media := extractMedia(item)
		if thumb != "" {
			e.Thumbnail = &thumb
		}
		if media != "" {
			e.MediaURL = &media
		}

		e.CreatedAt = time.Now().Unix()
		entries = append(entries, e)
	}
	return entries, nil
}

func itemGUID(item *gofeed.Item) string {
	if item.GUID != "" {
		return item.GUID
	}
	if item.Link != "" {
		return item.Link
	}
	return item.Title
}

func extractMedia(item *gofeed.Item) (thumbnail, mediaURL string) {
	if ext, ok := item.Extensions["media"]; ok {
		if contents, ok := ext["content"]; ok && len(contents) > 0 {
			if url := contents[0].Attrs["url"]; url != "" {
				mediaURL = url
			}
		}
		if thumbs, ok := ext["thumbnail"]; ok && len(thumbs) > 0 {
			if url := thumbs[0].Attrs["url"]; url != "" {
				thumbnail = url
			}
		}
	}

	if mediaURL == "" && len(item.Enclosures) > 0 {
		mediaURL = item.Enclosures[0].URL
	}

	if item.Image != nil && thumbnail == "" {
		thumbnail = item.Image.URL
	}

	return
}
